src/jsrs_parser.cc

```diff
@@ -318,32 +326,46 @@ MaybeLocal<Value> ParseNumber(Isolate*    isolate,
     }
   }
 
+  MaybeLocal<Value> result;
+
   if (base == 10) {
-    return ParseDecimalNumber(isolate, begin, end, size);
+    result = ParseDecimalNumber(isolate, number_start, end, size,
+                                negate_result);
   } else {
-    auto value = ParseIntegerNumber(isolate, number_start, end, size,
-                                    base, negate_result);
-    auto offset = static_cast<size_t>(number_start - begin);
-    *size += offset;
-    return value;
+    result = ParseIntegerNumber(isolate, number_start, end, size,
+                                base, negate_result);
   }
+  *size += number_start - begin;
+  return result;
 }
 
-Local<Value> ParseDecimalNumber(Isolate*    isolate,
-                                const char* begin,
-                                const char* end,
-                                size_t*     size) {
-  auto result = Number::New(isolate, atof(begin));
-  *size = end - begin;
-  size_t i = 0;
-  while (begin[i] != ',' &&
-         begin[i] != '}' &&
-         begin[i] != ']' &&
-         i < *size) {
-    i++;
+MaybeLocal<Value> ParseDecimalNumber(Isolate*    isolate,
+                                     const char* begin,
+                                     const char* end,
+                                     size_t*     size,
+                                     bool        negate_result) {
+  char* number_end;
+  double number = strtod(begin, &number_end);
+
+  if (negate_result) {
+    number = -number;
   }
-  *size = i;
-  return result;
+
+  // strictly allow only "NaN" and "Infinity"
+  if (isnan(number)) {
+    if (strncmp(begin + 1, "aN", 2) != 0) {
```

Comment:
What's so painful here? I just avoided testing for the same character twice here, that's it.